stan/math/rev/mat/fun/cholesky_decompose.hpp

```diff
@@ -0,0 +1,162 @@
+       * A(M_ - 1, M_ - 1), hence pos on line 94 is decremented
+       * to start at pos = M_ * (M_ + 1) / 2.
+       * */
+      virtual void chain() {
+        using Eigen::Matrix;
+        using Eigen::Dynamic;
+        using Eigen::RowMajor;
+        Matrix<double, Dynamic, Dynamic, RowMajor> adjL(M_, M_);
+        Matrix<double, Dynamic, Dynamic, RowMajor> LA(M_, M_);
+        Matrix<double, Dynamic, Dynamic, RowMajor> adjA(M_, M_);
+        size_t pos = 0;
+        for (size_type i = 0; i < M_; ++i)
+          for (size_type j = 0; j <= i; ++j) {
+            adjL.coeffRef(i, j) = _variRefL[pos]->adj_;
+            LA.coeffRef(i, j) = _variRefL[pos]->val_;
+            ++pos;
+          }
+
+        --pos;
+        for (int i = M_ - 1; i >= 0; --i) {
+          for (int j = i; j >= 0; --j) {
+            if (i == j) {
+              adjA.coeffRef(i, j) = 0.5 * adjL.coeffRef(i, j)
+                / LA.coeffRef(i, j);
+            } else {
+              adjA.coeffRef(i, j) = adjL.coeffRef(i, j)
+                / LA.coeffRef(j, j);
+              adjL.coeffRef(j, j) = adjL.coeffRef(j, j)
+                - adjL.coeffRef(i, j) * LA.coeffRef(i, j) / LA.coeffRef(j, j);
+            }
+            for (int k = j - 1; k >=0; --k) {
+              adjL.coeffRef(i, k) = adjL.coeffRef(i, k)
+                - adjA.coeffRef(i, j) * LA.coeffRef(j, k);
+              adjL.coeffRef(j, k) = adjL.coeffRef(j, k)
+                - adjA.coeffRef(i, j) * LA.coeffRef(i, k);
+            }
+            _variRefA[pos--]->adj_ += adjA.coeffRef(i, j);
+          }
+        }
+      }
+    };
+
+    /* Reverse mode specialization of
+     * cholesky decomposition
+     *
+     * Internally calls llt rather than using 
+     * stan::math::cholesky_decompose in order
+     * to use selfadjointView<Lower> optimization.
+     *
+     * Note chainable stack varis are created
+     * below in Matrix<var, -1, -1>
+     *
+     * @param Matrix A
+     * @return L cholesky factor of A
+     */
+    Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic>
+    cholesky_decompose(const Eigen::Matrix<var, -1, -1> &A) {
+      stan::math::check_square("cholesky_decompose", "A", A);
+      stan::math::check_symmetric("cholesky_decompose", "A", A);
+
+      Eigen::Matrix<double, -1, -1> L_A(value_of_rec(A));
+      Eigen::LLT<Eigen::MatrixXd> L_factor
+        = L_A.selfadjointView<Eigen::Lower>().llt();
+      check_pos_definite("cholesky_decompose", "m", L_factor);
+      L_A = L_factor.matrixL();
+
+      // NOTE: this is not a memory leak, this vari is used in the
+      // expression graph to evaluate the adjoint, but is not needed
+      // for the returned matrix.  Memory will be cleaned up with the
+      // arena allocator.
+      cholesky_decompose_v_vari *baseVari
+        = new cholesky_decompose_v_vari(A, L_A);
+
+      Eigen::Matrix<var, -1, -1> L(A.rows(), A.cols());
+      size_t pos = 0;
+      for (size_type i = 0; i < L.cols(); ++i) {
```

Comment:
It's all about memory locality.  What Eigen will do
internally is transpose a column-major matrix with all the necessary
memory allocation and copying because it's faster if you have to
reuse the rows.  So basically if the matrix is big enough that the
whole thing won't fit in cache.  

> On Aug 15, 2015, at 11:27 AM, Rob Trangucci <notifications@github.com> wrote:
> 
> In stan/math/rev/mat/fun/cholesky_decompose.hpp:
> 
> > +      Eigen::Matrix<double, -1, -1> L_A(value_of_rec(A));
> > +      Eigen::LLT<Eigen::MatrixXd> L_factor
> > +        = L_A.selfadjointView<Eigen::Lower>().llt();
> > +      check_pos_definite("cholesky_decompose", "m", L_factor);
> > +      L_A = L_factor.matrixL();
> > +
> > +      // NOTE: this is not a memory leak, this vari is used in the
> > +      // expression graph to evaluate the adjoint, but is not needed
> > +      // for the returned matrix.  Memory will be cleaned up with the
> > +      // arena allocator.
> > +      cholesky_decompose_v_vari *baseVari
> > +        = new cholesky_decompose_v_vari(A, L_A);
> > +
> > +      Eigen::Matrix<var, -1, -1> L(A.rows(), A.cols());
> > +      size_t pos = 0;
> > +      for (size_type i = 0; i < L.cols(); ++i) {
> 
> Now I'm not sure if this is a good idea. I don't know Eigen's internals well enough to determine what that'll do to computations done on L. What's the impact of multiplying a row-major storage order matrix by a column major storage order matrix? I'll run some tests to determine if there's any slowdown.
> 
> —
> Reply to this email directly or view it on GitHub.
>